### PR TITLE
Country grid

### PR DESCRIPTION
**Pull request by smikayel-opt**
### Desktop 
image
image
### Mobile
image
image

## Coverage
image
image

### CODE REVIEW

File: src/app/components/country-grid/country-grid.component.html [hunk 1]
```diff
@@ -1,5 +1,13 @@
-<app-loading *ngIf="loading"></app-loading>
-
-<div *ngIf="!loading">
-  here
+<div class="grid-container" [class.dark-theme]="theme === 'dark'">
```
Comment (hamzashamoun) [inline src/app/components/country-grid/country-grid.component.html]: toggling class in this way is correct, but try to use ngClass directive it's easier to handle multiple classes toggle in the future.

File: src/app/servicies/theme/theme.service.ts [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+export enum Theme {
+  LIGHT = 'light',
+  DARK = 'dark'
+}
+
+@Injectable({
+  providedIn: 'root'
+})
+export class ThemeService {
+  private themeSource = new BehaviorSubject<Theme>(Theme.LIGHT);
```
Comment (hamzashamoun) [inline src/app/servicies/theme/theme.service.ts]: you can subscribe to it directly in the other components without creating a new observable.

File: src/app/servicies/theme/theme.service.ts [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+  theme$ = this.themeSource.asObservable();
+
+  changeTheme(theme: Theme) {
```
Comment (hamzashamoun) [inline src/app/servicies/theme/theme.service.ts:19]: docs

File: src/app/servicies/theme/theme.service.spec.ts [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+  });
+
+  describe('theme$', () => {
+    it('should emit theme changes through the observable', (done: DoneFn) => {
+      const newTheme = Theme.DARK;
+
+      themeService.changeTheme(newTheme);
+      themeService.theme$.subscribe((theme) => {
+        expect(theme).toBe(newTheme);
+        done();
```
Comment (hamzashamoun) [inline src/app/servicies/theme/theme.service.spec.ts]: what is this function supposed to do?

File: src/app/servicies/country/country-api.service.ts [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+  * api for fetching the data about all countries
+  */
+  all(): Observable<ICountry[]> {
+    const query = 'all'
+    return this.http.get<ICountry[]>(this.API_URL + query)
+  }
+
+  /*
+  * search by region name like "Asia"
+  */
```
Comment (hamzashamoun) [inline src/app/servicies/country/country-api.service.ts:27]: missing region @param, add Semicolons please

File: src/app/servicies/country/country-api.service.ts [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+  searchByRegion(region: string): Observable<ICountry[]> {
+    const query = `region/${region}`
+    return this.http.get<ICountry[]>(this.API_URL + query)
+  }
+
+  /*
+  * search by country name like "Armenia"
+  */
```
Comment (hamzashamoun) [inline src/app/servicies/country/country-api.service.ts]: missing name @param, add Semicolons please

File: src/app/servicies/country/country-api.service.ts [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+})
+export class CountryApiService {
+  public API_URL = 'https://restcountries.com/v2/'
+
+  constructor(public http: HttpClient) { }
+
+  /*
+  * api for fetching the data about all countries
+  */
+  all(): Observable<ICountry[]> {
```
Comment (hamzashamoun) [inline src/app/servicies/country/country-api.service.ts]: please change the name for this method to be more obvious. ex: getAllCountries.
add Semicolons, please.

File: src/app/servicies/country/country-api.service.spec.ts [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+
+export const mockCountryData = [{ "name": "Armenia", "topLevelDomain": [".am"], "alpha2Code": "AM", "alpha3Code": "ARM", "callingCodes": ["374"], "capital": "Yerevan", "altSpellings": ["AM", "Hayastan", "Republic of Armenia", "Հայաստանի Հանրապետություն"], "subregion": "Western Asia", "region": "Asia", "population": 2963234, "latlng": [40.0, 45.0], "demonym": "Armenian", "area": 29743.0, "gini": 29.9, "timezones": ["UTC+04:00"], "borders": ["AZE", "GEO", "IRN", "TUR"], "nativeName": "Հայաստան", "numericCode": "051", "flags": { "svg": "https://flagcdn.com/am.svg", "png": "https://flagcdn.com/w320/am.png" }, "currencies": [{ "code": "AMD", "name": "Armenian dram", "symbol": "֏" }], "languages": [{ "iso639_1": "hy", "iso639_2": "hye", "name": "Armenian", "nativeName": "Հայերեն" }], "translations": { "br": "Armenia", "pt": "Arménia", "nl": "Armenië", "hr": "Armenija", "fa": "ارمنستان", "de": "Armenien", "es": "Armenia", "fr": "Arménie", "ja": "アルメニア", "it": "Armenia", "hu": "Örményország" }, "flag": "https://flagcdn.com/am.svg", "regionalBlocs": [{ "acronym": "EEU", "name": "Eurasian Economic Union", "otherAcronyms": ["EAEU"] }], "cioc": "ARM", "independent": true }]
```
Comment (hamzashamoun) [inline src/app/servicies/country/country-api.service.spec.ts]: too much data, can you please just put the data that you want to test

File: src/app/servicies/country/country-api.service.spec.ts [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+
+  it('should be created', () => {
+    expect(countryApiService).toBeTruthy();
+  });
+
+  describe('all', () => {
+    it('should return an Observable<ICountry[]> with all countries data', () => {
+      mockHttp.get.and.returnValue(of(mockCountryData));
+
+      countryApiService.all().subscribe((countries) => {
```
Comment (hamzashamoun) [inline src/app/servicies/country/country-api.service.spec.ts]: missing type for countries

File: src/app/servicies/country/country-api.service.spec.ts [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+      expect(mockHttp.get).toHaveBeenCalledWith(`${countryApiService.API_URL}all`);
+    });
+  });
+
+  describe('searchByRegion', () => {
+    it('should return an Observable<ICountry[]> with countries in the specified region', () => {
+      const mockRegion = 'Europe';
+      mockHttp.get.and.returnValue(of(mockCountryData));
+
+      countryApiService.searchByRegion(mockRegion).subscribe((countries) => {
```
Comment (hamzashamoun) [inline src/app/servicies/country/country-api.service.spec.ts]: missing type for countries

File: src/app/servicies/country/country-api.service.spec.ts [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+      expect(mockHttp.get).toHaveBeenCalledWith(`${countryApiService.API_URL}region/${mockRegion}`);
+    });
+  });
+
+  describe('searchByName', () => {
+    it('should return an Observable<ICountry[]> with countries matching the specified name', () => {
+      const mockName = 'Germany';
+      mockHttp.get.and.returnValue(of(mockCountryData));
+
+      countryApiService.searchByName(mockName).subscribe((countries) => {
```
Comment (hamzashamoun) [inline src/app/servicies/country/country-api.service.spec.ts]: missing type for countries

File: src/app/servicies/country/country-api.service.mock.ts [hunk 12]
```diff
@@ -0,0 +1,16 @@
+import { CountryApiService } from "./country-api.service";
+
```
Comment (hamzashamoun) [inline src/app/servicies/country/country-api.service.mock.ts]: remove extra line

File: src/app/components/select/select.component.ts [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+@Component({
+  selector: 'app-select',
+  standalone: true,
+  imports: [FormsModule, CommonModule],
+  templateUrl: './select.component.html',
+  styleUrl: './select.component.css'
+})
+export class SelectComponent {
+  @Input() selectedOption: string = '';
+  @Input() selectOptions: ISelect[] = []
```
Comment (hamzashamoun) [inline src/app/components/select/select.component.ts]: missing semicolon

File: src/app/components/select/select.component.spec.ts [hunk 14]
```diff
@@ -0,0 +1,36 @@
+import { SelectComponent } from './select.component';
+import { Theme, ThemeService } from '../../servicies/theme/theme.service';
+
+describe('SelectComponent', () => {
+  let component: SelectComponent;
+
+  beforeEach(() => {
+    component = new SelectComponent(new ThemeService());
+
```
Comment (hamzashamoun) [inline src/app/components/select/select.component.spec.ts]: extra line

File: src/app/components/select/select.component.spec.ts [hunk 15]
```diff
@@ -0,0 +1,36 @@
+import { SelectComponent } from './select.component';
+import { Theme, ThemeService } from '../../servicies/theme/theme.service';
+
+describe('SelectComponent', () => {
+  let component: SelectComponent;
+
+  beforeEach(() => {
+    component = new SelectComponent(new ThemeService());
```
Comment (hamzashamoun) [inline src/app/components/select/select.component.spec.ts]: create a mock for the ThemeService and then use it

File: src/app/components/search-input/search-input.component.spec.ts [hunk 16]
```diff
@@ -0,0 +1,35 @@
+import { ComponentFixture, TestBed } from '@angular/core/testing';
```
Comment (hamzashamoun) [inline src/app/components/search-input/search-input.component.spec.ts]: no TestBed please

File: src/app/components/search-input/search-input.component.ts [hunk 17]
```diff
@@ -0,0 +1,29 @@
...
+
+@Component({
+  selector: 'app-search-input',
+  standalone: true,
+  imports: [FormsModule],
+  templateUrl: './search-input.component.html',
+  styleUrl: './search-input.component.css'
+})
+export class SearchInputComponent {
+  value: string = ''
```
Comment (hamzashamoun) [inline src/app/components/search-input/search-input.component.ts]: missing semicolons

File: src/app/components/result-card/result-card.component.spec.ts [hunk 18]
```diff
@@ -0,0 +1,25 @@
...
+
+  describe('ngOnInit', () => {
+    it('should set the theme based on the value from themeService.getTheme', () => {
+      const expectedTheme = Theme.LIGHT;
+
+      component.ngOnInit();
+      expect(component.theme).toBe(expectedTheme);
+    });
+  });
+
```
Comment (hamzashamoun) [inline src/app/components/result-card/result-card.component.spec.ts]: extra line

File: src/app/components/result-card/result-card.component.spec.ts [hunk 19]
```diff
@@ -0,0 +1,25 @@
+import { ResultCardComponent } from './result-card.component';
+import { Theme, ThemeService } from '../../servicies/theme/theme.service';
+
+describe('ResultCardComponent', () => {
+  let component: ResultCardComponent;
+
+  beforeEach(() => {
+    component = new ResultCardComponent(new ThemeService());
```
Comment (hamzashamoun) [inline src/app/components/result-card/result-card.component.spec.ts]: use the mock Teme service

File: src/app/components/result-card/result-card.component.ts [hunk 20]
```diff
@@ -0,0 +1,25 @@
...
+  styleUrl: './result-card.component.css'
+})
+export class ResultCardComponent {
+  @Input() name: string = '';
+  @Input() flag: string = '';
+  @Input() population: number = 0;
+  @Input() region: string = '';
+  @Input() capital: string = '';
+
+  theme: Theme = Theme.LIGHT
```
Comment (hamzashamoun) [inline src/app/components/result-card/result-card.component.ts]: missing semicolon

File: src/app/components/header/header.component.spec.ts [hunk 21]
```diff
@@ -0,0 +1,41 @@
+import { HeaderComponent } from './header.component';
+import { Theme, ThemeService } from '../../servicies/theme/theme.service';
+
+describe('HeaderComponent', () => {
+  let component: HeaderComponent;
+
+  beforeEach(() => {
+    component = new HeaderComponent(new ThemeService());
```
Comment (hamzashamoun) [inline src/app/components/header/header.component.spec.ts]: use the mock ThemeService

File: src/app/components/header/header.component.ts [hunk 22]
```diff
@@ -0,0 +1,28 @@
...
+export class HeaderComponent {
+  theme: Theme = Theme.LIGHT
+
+  constructor(private themeService: ThemeService) { }
+
+  ngOnInit() {
+    this.theme = this.themeService.getTheme()
+  }
+
+  switchMode() {
```
Comment (hamzashamoun) [inline src/app/components/header/header.component.ts]: you can just change the value of the theme inside ThemeService directly and also use that value directly in all of the HTMLs components.
by doing that you don't need to create a new variable for each component and then change its value to the value that already exists in ThemeService.

File: src/app/components/filter/filter.component.ts [hunk 23]
```diff
@@ -0,0 +1,40 @@
...
+
+  selectOptions: ISelect[] = [
+    { title: "Filter By Region", value: "" },
+    { title: "Africa", value: "Africa" },
+    { title: "Asia", value: "Asia" },
+    { title: "Europe", value: "Europe" },
+    { title: "Oceania", value: "Oceania" },
+  ]
+
+  /**
```
Comment (hamzashamoun) [inline src/app/components/filter/filter.component.ts:25]: docs

File: src/app/components/filter/filter.component.ts [hunk 24]
```diff
@@ -0,0 +1,40 @@
...
+  * @param value 
+  */
+  onSearchQueryChanged(value: string) {
+    this.onCountryChange.emit(value)
+  }
+
+  /**
```
Comment (hamzashamoun) [inline src/app/components/filter/filter.component.ts:33]: docs

File: src/app/components/country-grid/country-grid.component.ts [hunk 25]
```diff
@@ -1,29 +1,79 @@
...
+   */
+  filterByName(countryName: string) {
+
+    if (!countryName) {
+      this.getAllCountries()
+      return
+    }
+
+    this.countryService.searchByName(countryName).subscribe(
+      (contries) => {
```
Comment (hamzashamoun) [inline src/app/components/country-grid/country-grid.component.ts]: missing semicolons and types for countries, error. we usually put type any for the errors

File: src/app/components/country-grid/country-grid.component.ts [hunk 26]
```diff
@@ -1,29 +1,79 @@
...
+   */
+  getAllCountries(): void {
+    this.countryService.all().subscribe(contries => {
+      this.countries = contries
     })
   }
+
+  /**
+   * get filltered countries , filtered by country name
+   * @param countryName
```
Comment (hamzashamoun) [inline src/app/components/country-grid/country-grid.component.ts]: missing param doc

File: src/app/components/country-grid/country-grid.component.ts [hunk 27]
```diff
@@ -1,29 +1,79 @@
...
-  ngOnInit() {
-    this.loading = true
-    this.countriesSerivce.all().subscribe(countriesData => {
-      this.countries = countriesData      
-      this.loading = false
+  /**
+   * get all countries from the country service
+   */
+  getAllCountries(): void {
+    this.countryService.all().subscribe(contries => {
```
Comment (hamzashamoun) [inline src/app/components/country-grid/country-grid.component.ts]: missing type and semicolons

File: src/app/components/country-grid/country-grid.component.ts [hunk 28]
```diff
@@ -1,29 +1,79 @@
...
+        this.countries = contries
+      },
+      (error) => {
+        this.countries = []
+      })
+  }
+
+  /**
+   * get filltered countries , filtered by region
+   * @param region
```
Comment (hamzashamoun) [inline src/app/components/country-grid/country-grid.component.ts]: missing semicolons and types for countries, error. we usually put type any for the errors
also missing param doc

File: src/app/components/country-grid/country-grid.component.spec.ts [hunk 29]
```diff
@@ -1,23 +1,83 @@
...
-    })
-    .compileComponents();
-    
-    fixture = TestBed.createComponent(CountryGridComponent);
-    component = fixture.componentInstance;
-    fixture.detectChanges();
+  const countryApiServiceMock = jasmine.createSpyObj('mockCountryApiService', ['all', 'searchByRegion', 'searchByName']);
+
+  beforeEach(() => {
+    component = new CountryGridComponent(new ThemeService(), countryApiServiceMock);
```
Comment (hamzashamoun) [inline src/app/components/country-grid/country-grid.component.spec.ts]: use mock for the ThemeService

**Review by hamzashamoun — CHANGES_REQUESTED** [review]

File: src/app/servicies/theme/theme.service.spec.ts (hunk 4 above)
Comment (smikayel-opt) [reply src/app/servicies/theme/theme.service.spec.ts]: I thought that it should check whether the theme can be changed with this function and return the expected new theme. However, after removing this test, it didn't change anything in coverage. It seems it's useless.I will remove the case

File: src/app/components/country-grid/country-grid.component.html [hunk 30]
```diff
@@ -1,8 +1,26 @@
...
+    (onSelectChange)="filterByRegion($event)"
+  ></app-filter>
+  <div class="pagination-container">
+    <button
+      [disabled]="currentPage === 1"
+      (click)="onPageChange(currentPage - 1)"
+    >
+      Previous
+    </button>
+    <span>Page {{ currentPage }} of {{ getTotalPages() }}</span>
```
Comment (hamzashamoun) [inline src/app/components/country-grid/country-grid.component.html]: I would like to see clickable numbers, that will take the user to the specific pages.
for example if the user clicks on number 3 the third page should appear

File: src/app/components/country-grid/country-grid.component.ts [hunk 31]
```diff
@@ -14,66 +14,114 @@ import { Theme, ThemeService } from '../../servicies/theme/theme.service';
...
    */
-  filterByName(countryName: string) {
+  updateDisplayedCountries(): void {
+    const startIndex = (this.currentPage - 1) * this.itemsPerPage;
+    const endIndex = startIndex + this.itemsPerPage;
+    this.displayedCountries = this.countries.slice(startIndex, endIndex);
+  }
+
+  /**
+   *
```
Comment (hamzashamoun) [inline src/app/components/country-grid/country-grid.component.ts]: missing method doc

File: src/app/components/country-grid/country-grid.component.ts [hunk 32]
```diff
@@ -14,66 +14,114 @@ import { Theme, ThemeService } from '../../servicies/theme/theme.service';
...
+   * @param searchCountryName country name from the input value
+   */
+  onSearchInputChange(searchCountryName: string): void {
+    this.currentPage = 1;
+    this.searchCountryName = searchCountryName;
+    this.filterByName(this.searchCountryName);
+  }
 
+  /**
+   *
```
Comment (hamzashamoun) [inline src/app/components/country-grid/country-grid.component.ts]: missing method doc

File: src/app/components/country-grid/country-grid.component.ts [hunk 33]
```diff
@@ -14,66 +14,114 @@ import { Theme, ThemeService } from '../../servicies/theme/theme.service';
...
       (error) => {
-        this.countries = []
-      })
+        this.countries = [];
+        this.updateDisplayedCountries();
+      });
   }
 
   /**
-   * get filltered countries , filtered by region
```
Comment (hamzashamoun) [inline src/app/components/country-grid/country-grid.component.ts]: please don't remove it 😭

File: src/app/components/country-grid/country-grid.component.ts [hunk 34]
```diff
@@ -14,66 +14,114 @@ import { Theme, ThemeService } from '../../servicies/theme/theme.service';
...
+      this.getAllCountries();
+      return;
     }
 
     this.countryService.searchByRegion(region).subscribe(
-      (contries) => {
-        this.countries = contries
+      (countries) => {
+        this.countries = countries;
+        console.log(countries)
```
Comment (hamzashamoun) [inline src/app/components/country-grid/country-grid.component.ts]: remove console.log

File: src/app/components/country-grid/country-grid.component.ts [hunk 35]
```diff
@@ -14,66 +14,114 @@ import { Theme, ThemeService } from '../../servicies/theme/theme.service';
...
-        this.countries = []
+        this.countries = [];
+        this.updateDisplayedCountries();
       }
-    )
+    );
+  }
+
+  /**
+   *
```
Comment (hamzashamoun) [inline src/app/components/country-grid/country-grid.component.ts]: missing method doc

File: src/app/servicies/theme/theme.service.ts [hunk 36]
```diff
@@ -0,0 +1,30 @@
...
+  /**
+  * will change the theme source from dark to light.. or from the light to the dark
+  * @param theme
+  */
+  changeTheme(theme: Theme) {
+    this.themeSource.next(theme);
+  }
+
+  /**
+   *
```
Comment (hamzashamoun) [inline src/app/servicies/theme/theme.service.ts]: missing method doc, semicolons

File: src/app/servicies/modal.service.ts [hunk 37]
```diff
@@ -0,0 +1,9 @@
+import { Injectable } from '@angular/core';
```
Comment (hamzashamoun) [inline src/app/servicies/modal.service.ts]: I can't see any usage for this service, if you don't want to use it please remove it.

File: src/app/servicies/country/country-api.service.ts [hunk 38]
```diff
@@ -0,0 +1,42 @@
...
+   * api for fetching the data about all countries
+   * @returns 
+   */
+  getAllCountries(): Observable<ICountry[]> {
+    const query = 'all';
+    return this.http.get<ICountry[]>(this.API_URL + query);
+  }
+
+  /**
+   *
```
Comment (hamzashamoun) [inline src/app/servicies/country/country-api.service.ts]: missing method doc

File: src/app/servicies/country/country-api.service.ts [hunk 39]
```diff
@@ -0,0 +1,42 @@
...
+   * @param region region name
+   * @returns 
+  */
+  searchByRegion(region: string): Observable<ICountry[]> {
+    const query = `region/${region}`;
+    return this.http.get<ICountry[]>(this.API_URL + query);
+  }
+
+  /**
+   *
```
Comment (hamzashamoun) [inline src/app/servicies/country/country-api.service.ts]: missing method doc

File: src/app/servicies/country/country-api.service.mock.ts (hunk 12 above)
Comment (hamzashamoun) [reply src/app/servicies/country/country-api.service.mock.ts]: @smikayel-opt

File: src/app/components/select/select.component.spec.ts (hunk 14 above)
Comment (hamzashamoun) [reply src/app/components/select/select.component.spec.ts]: @smikayel-opt

File: src/app/components/search-input/search-input.component.ts [hunk 40]
```diff
@@ -0,0 +1,29 @@
...
+})
+export class SearchInputComponent {
+  value: string = '';
+  @Output() onChange = new EventEmitter<string>();
+  theme: Theme = Theme.LIGHT;
+
+  constructor(private themeService: ThemeService) { }
+
+  ngOnInit() {
+    this.themeService.themeSource.subscribe(theme => this.theme = theme);
```
Comment (hamzashamoun) [inline src/app/components/search-input/search-input.component.ts]: now you can use the themeSource value directly, no need for creating a new variable in each component that uses the service.

File: src/app/components/result-card/result-card.component.ts (hunk 20 above)
Comment (hamzashamoun) [reply src/app/components/result-card/result-card.component.ts]: @smikayel-opt

File: src/app/components/header/header.component.ts (hunk 22 above)
Comment (hamzashamoun) [reply src/app/components/header/header.component.ts]: @smikayel-opt

File: src/app/components/filter/filter.component.ts (hunk 24 above)
Comment (hamzashamoun) [reply src/app/components/filter/filter.component.ts:33]: @smikayel-opt

File: src/app/components/filter/filter.component.ts (hunk 23 above)
Comment (hamzashamoun) [reply src/app/components/filter/filter.component.ts:25]: @smikayel-opt

File: src/app/components/filter/filter.component.ts [hunk 41]
```diff
@@ -0,0 +1,39 @@
+import { Component, EventEmitter, Input, Output } from '@angular/core';
+import { FormsModule } from '@angular/forms';
+import { SearchInputComponent } from '../search-input/search-input.component';
+import { ISelect, SelectComponent } from '../select/select.component';
+
```
Comment (hamzashamoun) [inline src/app/components/filter/filter.component.ts:5]: remove extra line

**Review by hamzashamoun — CHANGES_REQUESTED** [review]

**Review by hamzashamoun — APPROVED** [review]
Great job Sasha

**Pull request merged**